float32 ringX
float32 ringY
float32 normalX
float32 normalY
